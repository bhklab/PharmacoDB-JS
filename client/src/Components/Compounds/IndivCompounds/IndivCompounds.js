import React from 'react';
import styled from 'styled-components';
import Layout from '../../Layout/Layout';

/**
 * Parent component for the individual compounds page.
 *
 * @component
 * @example
 *
 * return (
 *   <Compounds/>
 * )
 */
const IndivCompounds = () => (
  <Layout>
    hi
  </Layout>
);

export default IndivCompounds;
