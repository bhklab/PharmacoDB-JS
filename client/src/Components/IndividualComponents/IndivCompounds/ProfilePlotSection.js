/* eslint-disable radix */
/* eslint-disable no-nested-ternary */
import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import PropTypes from 'prop-types';
// import { getSingleCompoundExperimentsQuery } from '../../../queries/experiments';
import dataset_colors from '../../../styles/dataset_colors';
import Loading from '../../Utils/Loading';
// import DatasetHorizontalPlot from '../../Plots/DatasetHorizontalPlot';

/**
 * Profile plots of the individula compound page.
 *
 * @component
 * @example
 *
 * return (
 *   <ProfilePlotSection/>
 * )
 */
const ProfilePlotSection = (props) => {
  const { compound } = props;
  const { id, name } = compound;

  // const { loading, error, data } = useQuery(getSingleCompoundExperimentsQuery, {
  //   variables: { compoundId: id },
  // });

  // if (loading) {
  //   return <Loading />;
  // }
  // if (error) {
  //   return <p> Error! </p>;
  // }
  return (
    <>
      <div>Profile Plot1</div>
      <div>Profile Plot2</div>
    </>
  );
};

ProfilePlotSection.propTypes = {
  compound: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  }).isRequired,
};

export default ProfilePlotSection;
