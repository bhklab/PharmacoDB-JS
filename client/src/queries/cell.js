import { gql } from 'apollo-boost';

/**
 * @returns - Query returns the list of cell lines with it's id and name,
 * and also returns the tissue object for the particular cell line.
 */
const getCellLinesQuery = gql`
  query getAllCellLines {
    cell_lines(all: true) {
      id
      name
      tissue {
        id
        name
      }
    }
  }
`;

/**
 * @param { Number } cellId - cell id of the cell line to be queried.
 * @returns - all the information returns by the cell lines query
 * and synonym for the cell line in different datasets.
 */
const getCellLineQuery = gql`
  query getSingleCellLine($cellId: Int!) {
    cell_line(cellId: $cellId) {
      id
      name
      tissue {
        id
        name
      }
      synonyms {
        name
        source
      }
    }
  }
`;

export {
  getCellLinesQuery,
  getCellLineQuery,
};
