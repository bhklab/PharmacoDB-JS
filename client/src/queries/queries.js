import { gql } from 'apollo-boost';

/**
 * Query returns the list of compounds with the id and name.
 */
const getCompoundsQuery = gql`
    {
        compounds {
            id
            name
        }
    }
`;

/**
 * Query returns info of compound given the id.
 */
const getCompoundQuery = gql`
 query Compound($compoundId: Int!) {
    compound(compoundId: $compoundId) {
      id
      name
    }
  }
`;

const disableDrug = gql`
  {
    drug @client
  }
`;

export {
  getCompoundsQuery,
  getCompoundQuery,
  disableDrug,
};
