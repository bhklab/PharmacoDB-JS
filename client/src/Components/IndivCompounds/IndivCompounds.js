/* eslint-disable no-nested-ternary */
import React from 'react';
import styled from 'styled-components';
import { useQuery } from '@apollo/react-hooks';
import Layout from '../Layout/Layout';
import { getCompoundQuery } from '../../queries/queries';

/**
 * Parent component for the individual compounds page.
 *
 * @component
 * @example
 *
 * return (
 *   <Compounds/>
 * )
 */

const IndivCompounds = (props) => {
  const { match: { params } } = props;
  const { loading, error, data } = useQuery(getCompoundQuery, {
    variables: { compoundId: parseInt(params.id) },
  });
  console.log(loading, error, data);
  return (
    <Layout>
      {loading ? (<p>Loading...</p>)
        : (error ? (<p>Error!</p>)
          : (
            <div>hi</div>
          ))}
    </Layout>
  );
};

export default IndivCompounds;
