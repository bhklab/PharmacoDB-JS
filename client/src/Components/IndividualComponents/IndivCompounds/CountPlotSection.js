/* eslint-disable radix */
/* eslint-disable no-nested-ternary */
import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import PropTypes from 'prop-types';
import { getSingleCompoundExperimentsQuery } from '../../../queries/experiments';
import dataset_colors from '../../../styles/dataset_colors';
import Loading from '../../Utils/Loading';
import DatasetHorizontalPlot from '../../Plots/DatasetHorizontalPlot';
import ProfileCellLine from '../../Plots/ProfileCellLine';

/**
 * A helper function that processes data from the API to be subsequently loaded it into
 * cell line and tissue dataset horizontal plots
 * @param {Array} experiments - list of experiments for a given drug returned by the API
 * @returns - array of two items. Elements of the array are a list of data points for tissue and cell line plots respectively
 * Each data point contains name, count and color properties
 * @example
 * [[{name: "CTRPv2", count: 25, color: "#ccebc5"}], ... ]
 */
const generateCountPlotData = (experiments) => {
  const tissueObj = {};
  const cellLineObj = {};
  experiments.forEach((experiment) => {
    if (cellLineObj[experiment.dataset.name]) {
      cellLineObj[experiment.dataset.name].push(experiment.cell_line.id);
    } else {
      cellLineObj[experiment.dataset.name] = [experiment.cell_line.id];
    }

    if (tissueObj[experiment.dataset.name]) {
      tissueObj[experiment.dataset.name].push(experiment.tissue.id);
    } else {
      tissueObj[experiment.dataset.name] = [experiment.tissue.id];
    }
  });
  const tissueData = Object.entries(tissueObj).map((dataset, i) => ({
    name: dataset[0],
    count: [...new Set(dataset[1])].length,
    color: dataset_colors[i],
  }));
  const cellLineData = Object.entries(cellLineObj).map((dataset, i) => ({
    name: dataset[0],
    count: [...new Set(dataset[1])].length,
    color: dataset_colors[i],
  }));

  return [tissueData, cellLineData];
};
/**
 * Section that display count plots for the individula compound page.
 *
 * @component
 * @example
 *
 * return (
 *   <CountPlotSection/>
 * )
 */
const CountPlotSection = (props) => {
  const { compound } = props;
  const { id, name } = compound;

  const { loading, error, data } = useQuery(getSingleCompoundExperimentsQuery, {
    variables: { compoundId: id },
  });
  if (loading) {
    return <Loading />;
  }
  if (error) {
    return <p> Error! </p>;
  }
  const [tissuesData, cellLinesData] = generateCountPlotData(data.experiments);
  return (
    <>
      <DatasetHorizontalPlot
        data={cellLinesData}
        xaxis="# of cell lines"
        title={`Number of cell lines tested with ${name} (per dataset)`}
      />
      <DatasetHorizontalPlot
        data={tissuesData}
        xaxis="# of tissues"
        title={`Number of tissues tested with ${name} (per dataset)`}
      />
      <ProfileCellLine data={data.experiments} />
    </>
  );
};

CountPlotSection.propTypes = {
  compound: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  }).isRequired,
};

export default CountPlotSection;
