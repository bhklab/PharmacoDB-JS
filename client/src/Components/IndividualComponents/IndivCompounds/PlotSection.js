/* eslint-disable radix */
/* eslint-disable no-nested-ternary */
import React, { useMemo } from 'react';
import { useQuery } from '@apollo/react-hooks';
import PropTypes from 'prop-types';
import { getSingleCompoundExperimentsQuery } from '../../../queries/experiments';
import dataset_colors from '../../../styles/dataset_colors';
import Loading from '../../UtilComponents/Loading';
import { generateOptions } from '../../../utils/plotProcessing';
import DatasetHorizontalPlot from '../../Plots/DatasetHorizontalPlot';
import ProfileCellLine from '../../Plots/ProfileCellLine';
import ProfileTissue from '../../Plots/ProfileTissue';
import PlotsWrapper from '../../../styles/PlotsWrapper';

/**
 * A helper function that processes data from the API to be subsequently loaded it into
 * cell line and tissue dataset horizontal plots
 * @param {Array} experiments - list of experiments for a given drug returned by the API
 * @returns - array of two items. Elements of the array are a list of data points for tissue and cell line plots respectively
 * Each data point contains name, count and color properties
 * @example
 * [[{name: "CTRPv2", count: 25, color: "#ccebc5"}], ... ]
 */
const generateCountPlotData = (experiments) => {
  const tissueObj = {};
  const cellLineObj = {};
  experiments.forEach((experiment) => {
    if (cellLineObj[experiment.dataset.name]) {
      cellLineObj[experiment.dataset.name].push(experiment.cell_line.id);
    } else {
      cellLineObj[experiment.dataset.name] = [experiment.cell_line.id];
    }

    if (tissueObj[experiment.dataset.name]) {
      tissueObj[experiment.dataset.name].push(experiment.tissue.id);
    } else {
      tissueObj[experiment.dataset.name] = [experiment.tissue.id];
    }
  });
  const tissueData = Object.entries(tissueObj).map((dataset, i) => ({
    name: dataset[0],
    count: [...new Set(dataset[1])].length,
    color: dataset_colors[i],
  }));
  const cellLineData = Object.entries(cellLineObj).map((dataset, i) => ({
    name: dataset[0],
    count: [...new Set(dataset[1])].length,
    color: dataset_colors[i],
  }));

  return [tissueData, cellLineData];
};
/**
 * Section that display plots for the individula compound page.
 *
 * @component
 * @example
 *
 * returns (
 *   <PlotSection/>
 * )
 */
const PlotSection = (props) => {
  const { display, compound } = props;
  const { id, name } = compound;

  const { loading, error, data } = useQuery(getSingleCompoundExperimentsQuery, {
    variables: { compoundId: id },
  });

  const experimentalData = data ? data.experiments : [];
  // memoization of the plotData
  const [tissuesData, cellLinesData] = useMemo(() => generateCountPlotData(experimentalData), [experimentalData]);
  const [profileOptions, datasetOptions] = useMemo(() => generateOptions(experimentalData), [experimentalData]);

  if (loading) {
    return '';
  }
  if (error) {
    return <p> Error! </p>;
  }

  return (
    <>
      {
        display === 'barplots' ?
          loading ? <Loading />
          :
          <PlotsWrapper>
            <DatasetHorizontalPlot
              data={cellLinesData}
              xaxis="# of cell lines"
              title={`Number of cell lines tested with ${name} (per dataset)`}
            />
            <DatasetHorizontalPlot
              data={tissuesData}
              xaxis="# of tissues"
              title={`Number of tissues tested with ${name} (per dataset)`}
            />
          </PlotsWrapper>
        :
        ''
      }
      {
        display === 'aacCells' ?
          loading ? <Loading />
          :
          <ProfileCellLine
            compound={name}
            data={experimentalData}
            profileOptions={profileOptions}
            datasetOptions={datasetOptions}
          />
        :
        ''
      }
      {
        display === 'aacTissues' ?
          loading ? <Loading />
          :
          <ProfileTissue
            compound={name}
            data={experimentalData}
            profileOptions={profileOptions}
            datasetOptions={datasetOptions}
          />
        :
        ''
      }
    </>
  );
};

PlotSection.propTypes = {
  compound: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  }).isRequired,
};

export default PlotSection;
