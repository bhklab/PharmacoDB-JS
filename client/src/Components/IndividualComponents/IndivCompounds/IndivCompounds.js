/* eslint-disable radix */
/* eslint-disable no-nested-ternary */
import React, { useState, useEffect } from 'react';
import { useQuery } from '@apollo/react-hooks';
import { Element } from 'react-scroll';
import PropTypes from 'prop-types';
import Layout from '../../UtilComponents/Layout';
import { getCompoundQuery } from '../../../queries/compound';
import { NotFoundContent } from '../../UtilComponents/NotFoundPage';
import Table from '../../UtilComponents/Table/Table';
import PlotSection from './PlotSection';
import CellLinesSummaryTable from './Tables/CellLinesSummaryTable'
import TissuesSummaryTable from './Tables/TissuesSummaryTable'
import MolecularFeaturesTable from './Tables/MolecularFeaturesTable'
import {StyledIndivPage, StyledSidebarList} from '../../../styles/IndivPageStyles';
import StyledWrapper from '../../../styles/utils';

const SYNONYM_COLUMNS = [
    {
        Header: 'Sources',
        accessor: 'source',
        Cell: (item) => {
            let datasets = item.cell.row.original.source;
            return(datasets.map((obj, i) => (
                    obj.id? (
                            <span key={i}>
                        <a href={`/datasets/${obj.id}`}>{obj.name}</a>{ i + 1 < datasets.length ? ', ' : ''}
                    </span>
                        ) :
                        (<span key={i}>{obj.name}</span>)
                )
            ));
        }
    },
    {
        Header: 'Names Used',
        accessor: 'name',
    },
];

const ANNOTATION_COLUMNS = [
    {
        Header: 'Database',
        accessor: 'db',
    },
    {
        Header: 'Identifier',
        accessor: 'identifier',
    },
];

const SIDE_LINKS = [
    {label: 'Synonyms and IDs', name: 'synonyms'},
    {label: 'Bar Plots', name: 'barplots'},
    {label: 'AAC (Cell Lines)', name: 'aacCells'},
    {label: 'AAC (Tissues)', name: 'aacTissues'},
    {label: 'Cell Line Summary', name: 'cellSummary'},
    {label: 'Tissue Summary', name: 'tissueSummary'},
    {label: 'Molecular Features', name: 'molFeature'},
];

/**
 * Format data for the synonyms table
 * @param {Array} data synonym data from the experiment API
 */
const formatSynonymData = (data) => {
    if (data.synonyms) {
        const returnObj = data.synonyms;
        returnObj.push({name:data.compound.name , source:[{name: "PharmacoGx", id: ''}]})
        return returnObj;
    }
    return null;
};

/**
 * Format data for the external ids annotation table
 * @param {Array} data annotation data from the compound API
 */
const formatAnnotationData = (data) => {
    const modifiedData = [];
    if (data) {
        const { annotation } = data;
        if (annotation.smiles) {
            modifiedData.push({ db: 'SMILES', identifier: annotation.smiles, });
        }
        if (annotation.inchikey) {
            modifiedData.push({ db: 'InChiKey', identifier: annotation.inchikey, });
        }
        if (annotation.pubchem) {
            modifiedData.push({ db: 'PubChem ID', identifier: annotation.pubchem, });
        }
    }
    return modifiedData;
};

/**
 * Parent component for the individual compound page.
 *
 * @component
 * @example
 *
 * return (
 *   <IndivCompounds/>
 * )
 */
const IndivCompounds = (props) => {
    // parameter.
    const {
        match: { params },
    } = props;
    // query to get the data for the single compound.
    const { loading, error, data: queryData } = useQuery(getCompoundQuery, {
        variables: { compoundId: parseInt(params.id) },
    });

    // load data from query into state
    const [compound, setCompound] = useState({
        data: {},
        loaded: false,
    });
    // A section to display on the page
    const [display, setDisplay] = useState('synonyms');

    // to set the state on the change of the data.
    useEffect(() => {
        if (queryData !== undefined) {
            setCompound({
                data: queryData.singleCompound,
                loaded: true,
            });
        }
    }, [queryData]);

    // destructuring the compound object.
    const { data } = compound;

    // formatted data for synonyms annotation table
    const synonymColumns = React.useMemo(() => SYNONYM_COLUMNS, []);
    const synonymData = React.useMemo(() => formatSynonymData(data), [
        data.synonyms,
    ]);

    // formatted data for external ids annotation table
    const annotationColumns = React.useMemo(() => ANNOTATION_COLUMNS, []);
    const annotationData = React.useMemo(
        () => formatAnnotationData(data.compound),
        [data.compound]
    );

    /**
     *
     * @param {String} link
     */
    const createSideLink = (link, i) => (
        <li key={i} className={display === link.name ? 'selected': undefined}>
            <button type='button' onClick={() => setDisplay(link.name)}>
                {link.label}
            </button>
        </li>
    );
    return compound.loaded ? (
        <Layout page={data.compound.name}>
            <StyledWrapper>
                {loading ? (
                    <p>Loading...</p>
                ) : error ? (
                    <NotFoundContent />
                ) : (
                    <StyledIndivPage className="indiv-compounds">
                        <div className='heading'>
                            <span className='title'>{data.compound.name}</span>
                            <span className='attributes'>
                                FDA Approval Status:
                                <span className={`value ${data.compound.annotation.fda_status === 'Approved' ? 'highlight' : 'regular'}`}>
                                    {data.compound.annotation.fda_status}
                                </span>
                            </span>
                        </div>
                        <div className='wrapper'>
                            <StyledSidebarList>
                                {SIDE_LINKS.map((link, i) => createSideLink(link, i))}
                            </StyledSidebarList>
                            <div className="container">
                                <div className="content">
                                    {
                                        display === 'synonyms' &&
                                        <React.Fragment>
                                            <Element className="section" name="synonyms">
                                                <div className='section-title'>Synonyms</div>
                                                <Table
                                                    columns={SYNONYM_COLUMNS}
                                                    data={synonymData}
                                                    disablePagination
                                                />
                                            </Element>
                                            {
                                                annotationData.length > 0 ?
                                                    <Element
                                                        className="section"
                                                        name="external_ids"
                                                    >
                                                        <div className='section-title'>External IDs</div>
                                                        <Table
                                                            columns={annotationColumns}
                                                            data={annotationData}
                                                            disablePagination
                                                        />
                                                    </Element> :
                                                    ''
                                            }
                                        </React.Fragment>
                                    }
                                    {
                                        display === 'targets' &&
                                        <Element className="section" name="annotated_targets">
                                            <div className='section-title'>Annotated Targets</div>
                                            <div className="text">
                                                {data.targets
                                                    ? data.targets
                                                        .map((x) => x.name)
                                                        .join(', ')
                                                    : ''}
                                            </div>
                                        </Element>
                                    }
                                    <Element>
                                        <PlotSection
                                            display={display}
                                            compound={{
                                                id: data.compound.id,
                                                name: data.compound.name,
                                            }}
                                        />
                                    </Element>
                                    {
                                        display === 'cellSummary' &&
                                        <Element className="section">
                                            <div className='section-title'>Cell Line Summary</div>
                                            <CellLinesSummaryTable compound={({ id: data.compound.id, name: data.compound.name })}/>
                                        </Element>
                                    }
                                    {
                                        display === 'tissueSummary' &&
                                        <Element className="section">
                                            <div className='section-title'>Tissue Summary</div>
                                            <TissuesSummaryTable compound={({ id: data.compound.id, name: data.compound.name })}/>
                                        </Element>
                                    }
                                    {
                                        display === 'molFeature' &&
                                        <Element className="section">
                                            <div className='section-title'>Molecular Features</div>
                                            <MolecularFeaturesTable compound={({ id: data.compound.id, name: data.compound.name })}/>
                                        </Element>
                                    }
                                </div>
                            </div>
                        </div>
                    </StyledIndivPage>
                )}
            </StyledWrapper>
        </Layout>
    ) : null;
};

IndivCompounds.propTypes = {
    /**
     * IndivCompounds' param id
     */
    match: PropTypes.shape({
        params: PropTypes.shape({
            id: PropTypes.string.isRequired,
        }).isRequired,
    }).isRequired,
};

export default IndivCompounds;
