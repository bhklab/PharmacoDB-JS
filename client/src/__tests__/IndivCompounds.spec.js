import React from 'react';
import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { wait, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
// for async tests
import regeneratorRuntime from 'regenerator-runtime';
import IndivCompounds from '../Components/IndivCompounds/IndivCompounds';
import { getCompoundQuery } from '../queries/queries';

const mocks = [
  {
    request: {
      query: getCompoundQuery,
      variables: {
        compoundId: 1,
      },
    },
    result: {
      data: {
        compound: {
          compound: {
            id: '1',
            name: 'testDrug',
            annotation: {
              smiles: 'OCO',
              inchikey: '0',
              pubchem: '0',
            },
          },
        },
      },
    },
  },
];

// it('renders correctly with a mocked query and no id prop', () => {
//   // must use mount to access IndivCompounds (because child component)
//   const component = mount(
//     <MockedProvider mocks={mocks} addTypename={false}>
//       <IndivCompounds />
//     </MockedProvider>,
//   );
//   expect(component).toMatchSnapshot();
// });

it('renders correctly with a mocked query and id prop', async () => {
  // must use mount to access IndivCompounds (because child component)
  const component = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <IndivCompounds match={{ params: { id: '1' } }} />
      </BrowserRouter>
    </MockedProvider>,
  );
  await act(wait);
  expect(component).toMatchSnapshot();
});
