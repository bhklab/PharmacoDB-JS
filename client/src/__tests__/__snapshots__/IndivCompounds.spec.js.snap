// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndivCompounds renders correctly with a mocked query and a valid id prop 1`] = `Promise {}`;

exports[`IndivCompounds renders correctly with a mocked query and an invalid id prop (ex. null) 1`] = `Promise {}`;
